csrc/type.h

```diff
@@ -95,16 +95,34 @@ struct PointerOf {
   inline bool operator==(const PointerOf& other) const;
 };
 
+struct StructOf {
+  // In theory, we should just use std::unordered_map<std::string, T>, but this
+  // doesn't work on old gcc. See also SetTheoreticNaturalNumbers
+#if defined(__clang__) || __GNUC__ >= 12
+  std::unordered_map<std::string, DataType> types;
+#define GCC_BUG(x) x
```

Comment:
I agree with the `NVFUSER_` prefix and yes, it is exposed globally. Another information:

I did some searching on STL containers with incomplete type. The summary is: it is totally possible to implement STL containers supporting incomplete type (actually, boost has these container implementations), and it totally make sense to implement STL containers that way. However, due to historical reason, some standard C++ libraries are not implementing STL containers that way, and after careful consideration, the C++ standard committee decided to not write the requirement on supporting incomplete type into the standard because they didn't want to deprecate these libraries. However, starting from C++17, the standard start to ask `std::vector`, `std::list` and `std::forward_list` to support incomplete type. So

```C++
struct A;
std::vector<A> a; // valid on C++17
std::unordered_set<A> s; // undefined behavior, should be working on newer gcc.
struct A {};
``` 

So I would not blame this as a "gcc bug", and I will choose a better name for this macro, and put this message into the code as comment. But still, for cases where we know that `std::unordered_set` supports incomplete type, I still prefer to use it with a `#if` guard to enable it, because this should be the direction of C++ standard.